File: clojure/src/project_euler/sol/p0019.clj
(ns project-euler.sol.p0019)

(def ^:private days
  ;; days in month (Jan 1901 - *Nov* 2000)
  (let [common-year '(31 28 31 30 31 30 31 31 30 31 30 31)
        leap-year '(31 29 31 30 31 30 31 31 30 31 30 31)]
    (drop-last (flatten (repeat 25 (concat (repeat 3 common-year) leap-year))))))

(defn solve
  []
  ;; day of week - [0: Sunday, 1: Monday, 2: Tuesday, ..., 6: Saturday]
  ;; Jan 1, 1900 was Monday.
  (->> (reductions + (+ 1 365) days)
       (filter #(zero? (mod % 7)))
       (count)))

File: haskell/src/Sol/P0019.hs
module Sol.P0019 (compute, solve) where

import Mylib.Util (initExn)

compute :: String
compute =
    show
        . length
        . filter (== 6) -- '6' indicates Sunday because Jan 1, 1900 was a Monday.
        $ scanl (\acc x -> (acc + x) `mod` 7) 0 (365 : days) -- 365 is the number of days in 1900
  where
    -- the number of days in each month (Jan 1901 - Nov 2000)
    days :: [Int]
    days =
        initExn -- drop 'Dec 2000'
            . concat -- from 'Jan 1901' to 'Dec 2000'
            $ replicate 25 (commonYear ++ commonYear ++ commonYear ++ leapYear)
    commonYear = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    leapYear = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

solve :: String
solve = compute
